part2/decomposition.py

```
@@ -73,10 +73,12 @@ def jacobi_eigenvalues(
             V[i][q] = s * v_ip + c * v_iq
 
     if not is_converged:
-        raise ValueError(
-            f"Numerical instability: Failed to converge within {max_iterations} iterations"
+        import warnings
+        warnings.warn(
```

Ruff (0.15.9)

[warning] No explicit `stacklevel` keyword argument found

Set `stacklevel=2`

(B028)